Document breaking change in crypto introduced in 3.0 (#20366)
dotnet/docs: includes/core-changes/cryptography/3.0/cryptography-cryptostream-dispose-final-block-write.md

> ### System.Security.Cryptography.CryptoStream.Dispose is transforming final block only when writing

Reviewer: ```suggestion
### CryptoStream.Dispose transforms final block only when writing
```

You'll also need to update the bookmark links.

> The <xref:System.Security.Cryptography.CryptoStream.Dispose?displayProperty=fullName> which is used to finish crypto stream operations was previously throwing an exception when used with some crypto transforms (i.e. AES with padding) and reading was not complete. This scenario can occur i.e. in network scenarios when operation is cancelled and causing users to have to catch the exception.

Reviewer 2: https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/i/ie

Reviewer 3: I would just drop the "i.e." from that sentence.

> The <xref:System.Security.Cryptography.CryptoStream.Dispose?displayProperty=fullName> which is used to finish crypto stream operations was previously throwing an exception when used with some crypto transforms (i.e. AES with padding) and reading was not complete. This scenario can occur i.e. in network scenarios when operation is cancelled and causing users to have to catch the exception.

Reviewer 4: ```suggestion
The <xref:System.Security.Cryptography.CryptoStream.Dispose?displayProperty=fullName> method, which is used to finish CryptoStream operations, was previously throwing an exception when used with some crypto transforms (i.e. AES with padding) and reading was not complete. This scenario can occur i.e. in network scenarios when operation is cancelled and causing users to have to catch the exception.
```

(inserted "method" between "Dispose" and "which"; moved which ... operations to an aside, changed "crypto stream" to "CryptoStream")

> The <xref:System.Security.Cryptography.CryptoStream.Dispose?displayProperty=fullName> which is used to finish crypto stream operations was previously throwing an exception when used with some crypto transforms (i.e. AES with padding) and reading was not complete. This scenario can occur i.e. in network scenarios when operation is cancelled and causing users to have to catch the exception.

Reviewer: ```suggestion
The <xref:System.Security.Cryptography.CryptoStream.Dispose%2A?displayProperty=nameWithType> method, which is used to finish `CryptoStream` operations, no longer attempts to transform the final block when reading.
```

#### Change description

> In previous versions, when using <xref:System.Security.Cryptography.CryptoStream> in <xref:System.Security.Cryptography.CryptoStreamMode.Read> mode and user performed incomplete read the <xref:System.Security.Cryptography.CryptoStream.Dispose> could throw an exception (i.e. when using AES with padding) as final block was attempted to be transformed but data was incomplete.

Reviewer: ```suggestion
In previous .NET versions, if a user performed an incomplete read when using <xref:System.Security.Cryptography.CryptoStream> in <xref:System.Security.Cryptography.CryptoStreamMode.Read> mode, the <xref:System.Security.Cryptography.CryptoStream.Dispose%2A> method could throw an exception (for example, when using AES with padding). The exception was thrown because the final block was attempted to be transformed but the data was incomplete.
```

> In .NET 3.0 and later versions, the <xref:System.Security.Cryptography.CryptoStream.Dispose> will no longer try to transform the final block when reading which will allow for incomplete reads.

Reviewer: ```suggestion
In .NET 3.0 and later versions, <xref:System.Security.Cryptography.CryptoStream.Dispose%2A> no longer tries to transform the final block when reading, which allows for incomplete reads.
```

#### Reason for change

> This change enables incomplete reads from the crypto stream (i.e. when cancelling network operations) without need to catch an exception.

Reviewer: ```suggestion
This change enables incomplete reads from the crypto stream when a network operation is cancelled, without the need to catch an exception.
```

#### Version introduced

3.0

#### Recommended action

> Most apps should not be affected by this change. If application previously caught an exception in order to suppress exception caused by the incomplete read it will no longer be needed.

Reviewer: ```suggestion
Most apps should not be affected by this change.

If your application previously caught an exception in case of an incomplete read, you can delete that `catch` block.
```

> If app used transforming of the final block in hashing scenarios it might now need to ensure the entire stream is read before stream is disposed.

Reviewer: ```suggestion

If your app used transforming of the final block in hashing scenarios, you might need to ensure that the entire stream is read before it's disposed.
```

#### Category

Cryptography

#### Affected APIs

> - <xref:System.Security.Cryptography.CryptoStream.Dispose?displayProperty=fullName>

Reviewer 2: Build is reporting xref not found

Author: how do I ref it then?

Reviewer 2: Not sure what's the correct action here.
cc: @gewarren

Author: I've talked with @gewarren offline, apparently adding `%2A` (hex for `*`) after `Dispose` should do the trick

<!--

#### Affected APIs

> - `P:System.Security.Cryptography.CryptoStream.Dispose`

Reviewer 2: I think this should be "M" (method) not "P" (property).

### CryptoStream.Dispose transforms final block only when writing

The <xref:System.Security.Cryptography.CryptoStream.Dispose%2A?displayProperty=nameWithType> method, which is used to finish `CryptoStream` operations, no longer attempts to transform the final block when reading.

#### Change description

In previous .NET versions, if a user performed an incomplete read when using <xref:System.Security.Cryptography.CryptoStream> in <xref:System.Security.Cryptography.CryptoStreamMode.Read> mode, the <xref:System.Security.Cryptography.CryptoStream.Dispose%2A> method could throw an exception (for example, when using AES with padding). The exception was thrown because the final block was attempted to be transformed but the data was incomplete.

> In .NET 3.0 and later versions, <xref:System.Security.Cryptography.CryptoStream.Dispose%2A> no longer tries to transform the final block when reading, which allow sfor incomplete reads.

Reviewer 2: ```suggestion
In .NET Core 3.0 and later versions, <xref:System.Security.Cryptography.CryptoStream.Dispose%2A> no longer tries to transform the final block when reading, which allows for incomplete reads.
```

#### Reason for change

This change enables incomplete reads from the crypto stream when a network operation is cancelled, without the need to catch an exception.

#### Version introduced

3.0

#### Recommended action

Most apps should not be affected by this change.

If your application previously caught an exception in case of an incomplete read, you can delete that `catch` block.
If your app used transforming of the final block in hashing scenarios, you might need to ensure that the entire stream is read before it's disposed.

#### Category

Cryptography

#### Affected APIs

> - <xref:System.Security.Cryptography.CryptoStream.Dispose?displayProperty=fullName>

Reviewer 2: ```suggestion
- <xref:System.Security.Cryptography.CryptoStream.Dispose%2A?displayProperty=fullName>
```
